tools/test-runner/src/buck_test.rs

```diff
@@ -0,0 +1,252 @@
+            let mut err = String::new();
+            child.stderr.unwrap().read_to_string(&mut err).unwrap();
+            return TestResult {
+                name: test.name,
+                attempts,
+                passed: true,
+                duration,
+                stdout: out,
+                stderr: err,
+                contacts: test.contacts,
+            };
+        } else {
+            if attempts >= 1 + retries {
+                let mut out = String::new();
+                child.stdout.unwrap().read_to_string(&mut out).unwrap();
+                let mut err = String::new();
+                child.stderr.unwrap().read_to_string(&mut err).unwrap();
+                return TestResult {
+                    name: test.name,
+                    attempts,
+                    passed: false,
+                    duration,
+                    stdout: out,
+                    stderr: err,
+                    contacts: test.contacts,
+                };
+            }
+        }
+    }
+}
+
+// a.k.a., the defaults
+pub mod shell {
+    use super::{make_command, Test, TestKind, TestSpec};
+    use std::process::Child;
+
+    /// Builds a singleton test from a given spec.
+    pub fn validate(spec: TestSpec) -> Vec<Test> {
+        let command = make_command(
+            spec.command[0].clone(),
+            spec.command[1..].to_vec(),
+            spec.env,
+            spec.cwd,
+        );
+        let test = Test {
+            command,
+            name: spec.target,
+            labels: spec.labels,
+            contacts: spec.contacts,
+            kind: TestKind::Shell,
+        };
+        return vec![test]; // i.e; "just run it"
+    }
+
+    /// Validates the test based solely on its return value.
+    pub fn evaluate(result: &mut Child) -> bool {
+        result.wait().unwrap().success() // shouldn't block, it already ran
+    }
+}
+
+/// Refer to https://buck.build/files-and-dirs/buckconfig.html#test.external_runner
+///
+/// A JSON spec needs at least these fields, any others will be ignored.
+#[derive(Clone, Debug, Deserialize)]
+pub struct TestSpec {
+    /// The buck target URI of the test rule.
+    pub target: String,
+
+    /// The type of the test.
+    #[serde(rename(deserialize = "type"), default = "default_kind")]
+    kind: TestKind,
+
+    /// Command line that should be used to run the test.
+    pub command: Vec<String>,
+
+    /// Environment variables to be defined when running the test.
+    pub env: HashMap<String, String>,
+
+    /// Labels that are defined on the test rule.
+    pub labels: HashSet<String>,
+
+    /// Contacts that are defined on the test rule.
+    pub contacts: HashSet<String>,
+
+    /// Working directory the test should be run from.
+    pub cwd: Option<PathBuf>,
+
+    /// Absolute paths to any files required by this test target.
+    required_paths: Option<Vec<PathBuf>>,
+}
+fn default_kind() -> TestKind {
```

Comment:
Does this allow falling back to `shell` tests for the unsupported `gtest` type? If so cool, otherwise I still like it :)

Reply:
I don't think so. This default only applies when the JSON doesn't have a type tag at all.
If serde finds `"type" : <something>` in the test spec, `<something>` needs to match some case in the `TestKind` enum.

Reply:
Gotcha, makes sense

Reply:
After realizing buck does add a `"custom"` tag when no type is provided in `sh_test`, I removed this default.
In the future, we could change the parsing behavior so as to assign the `Shell` type to anything not recognized. This would mean implementing a custom serde [Deserialize](https://serde.rs/impl-deserialize.html) for the `TestKind` enum.